### ISSUE TITLE
Q3 Goals discussion

### ISSUE THREAD

**Issue by mattkime** [meetup/meetup-web-platform#302]
Proposed Q3 goals

- Improved production reliability
  - Blue / green deployment
  - Improved alerting / monitoring
    - Needs more detail
- Modularization of platform
  - Need to specify which modules will be broken out
- API Caching
  - Will require work with...
- Product team support
  - To be defined....

Objective
- Deploy faster with higher quality

Key Results
- 

Maybe 
- Performance improvements
  - browser preload
- Better modularization of platform and tools
- Live UI updating
- Offline first

Failed to make the cut
- HTTP/2 push (browsers not ready)

**Comment by mmcgahan** [comment]
Thanks for putting this up. I'd like to know more about some of the goals to get a better sense of how they relate to what we're currently doing and what you're thinking we would have to do to achieve them.

> 1. Offline-first

this seems like a good goal. I'd be cautious about setting it as a Q3 vs an H2 goal given that there's a fair amount of experimental work involved

> 2. API Caching:

- On the Hapi server or in the browser?
- Caching the REST API calls on the server should probably wait until we have upstream support for time-to-live headers (`Cache-Control`/`Expires`), which aren't currently implemented. I think there is going to be a caching effort in H2, but I think we should avoid setting up a dependency that we don't have a lot of control over.
- There are many different ways of approaching this problem, and I would hope that we could do a full survey of options in the course of a few months, but implementation would likely take longer

> 3. Live UI updating

I think this refers to part of the discussion in the PWA doc, but as I [mentioned there](https://github.com/meetup/meetup-web-platform/issues/274#issuecomment-304975471), we'd probably want to do this with Server Side Events, which is HTTP/2 dependent and not fully supported across browsers. We could potentially implement it with websockets, but my experience from Coco suggests that a websocket implementation would be months of work that would then most likely be scrapped for SSE eventually anyway.

The larger immediate obstacle which is REST API support - there would need to be a very large BE project dedicated to streaming data for each endpoint that we wanted to update live, and that's probably not going to happen this year. Also, my experience from Coco using streaming servers makes me very cautious about putting a timeframe around this support - I suspect it will need to be product-team driven through a coordinated BE + Web eng effort.

> 4. We could do another pass at alerting / monitoring

I'd like to know more specifics about this before putting a timeframe on it. We can and should put together a document describing our ideal alerting/monitoring behavior, but it's one of those things where it could take anywhere from 1 to 5 sprints to assemble different tools

> Blue / green deployment

My impression at the moment is that this will be partially dependent on other teams, which makes it hard to timebox. Definitely something we could work on, but I'm cautious about marking it as a goal, particularly if it ended up requiring significant changes to our deployment infrastructure.

---

Given that we've got the core features running reasonably well in production at the moment, the thing that I am most interested in doing now is consolidating the platform into a more coherent and well-defined library of modules - mainly described in #281. Essentially, I'd like to take the time to address the tech debt accumulated over the past year+ of experimentation and MVP-oriented releases. Now that we're in production, our standards for releases need to be even higher and we really should be polishing the system to be robust enough to adapt to the needs of the next 10 years of development. It's not currently in a state where I would be comfortable passing it on to a new developer and expecting them to use it effectively without a lot of guidance from the web platform team, but I think we need to get it to that state in order for it to be a long-term success.

This means slowing down the focus on 'features' like PWA, which will add complexity.

Implementing the platform as a well-defined set of modules will make _all future development_ simpler, as well as give us the opportunity to genuinely open-source some of the more polished modules that we write - we can take the time to properly document them and set them up as packages in npm, and make a genuine effort to share what we've learned with the wider community.

If we ended up at the end of the year with a handful of OSS packages that were well documented, tested, and maintained, I think we could consider it a successful H2 - caching and offline-first support don't seem as high-priority to me given the current prod metrics.

**Comment by mattkime** [comment]
>this seems like a good goal. I'd be cautious about setting it as a Q3 vs an H2 goal given that there's a fair amount of experimental work involved

Can you define a Q3 goal or goals?

> Caching - On the Hapi server or in the browser?

Seems like there's overlap here with offline first. 

>Caching the REST API calls on the server should probably wait until we have upstream support for time-to-live headers

Since we're creating the UI I think we have a chance to drive the conversation here. How fresh does info need to be for a given view?

I agree that server level caching is likely out for Q3 - at least the sort thats our responsibility. Other API consumers will want caching server side caching. If we imagine having a highly cached API server, how much time would we serve by having a mup-web local cache? 

>I think this refers to part of the discussion in the PWA doc, but as I mentioned there, we'd probably want to do this with Server Side Events

Then we're back to doing it with API polling. I'm not sure I understand your reasons against this. I think my worry at this point would be that we're trying to reduce server load and this wouldn't do that.

**Comment by mmcgahan** [comment]
> Can you define a Q3 goal or goals?

Sure - I did a writeup earlier this week that I'll share as a separate comment

>> Caching - On the Hapi server or in the browser?

> Seems like there's overlap here with offline first. 

Yes there is - we should be more specific about what is part of the offline-first caching strategy and what other strategies we would like to implement. Offline-first always uses a service worker, whereas the API requests can be cached in lots of different places. In the browser, we're currently caching API request/responses in IndexedDB using Redux middleware, but that doesn't simulate network responses the way a service worker would. 

>> Caching the REST API calls on the server should probably wait until we have upstream support for time-to-live headers

> Since we're creating the UI I think we have a chance to drive the conversation here. How fresh does info need to be for a given view?

We should definitely be part of the conversation - I expect that conversation will be ongoing until late in the year, and although we could do some preliminary development to support the expected BE implementation, I don't think we can count on having anything complete in Q3 and maybe not even H2.

> I agree that server level caching is likely out for Q3 - at least the sort thats our responsibility. Other API consumers will want caching server side caching. If we imagine having a highly cached API server, how much time would we serve by having a mup-web local cache?

So, there are 3 broad caching layers we're discussing:

1. the REST API
2. the Node API proxy endpoint
3. the browser

We don't have to worry about (1), and we would need to evaluate different options for (2). I think we will eventually end up with _some_ API caching in the Node layer because it would eliminate a ton of GCP-AWS http requests, which has both performance and cost benefits, but it's complex and dependent on a lot of upstream work so probably not going to be viable in the near term.

>> I think this refers to part of the discussion in the PWA doc, but as I mentioned there, we'd probably want to do this with Server Side Events

> Then we're back to doing it with API polling. I'm not sure I understand your reasons against this. I think my worry at this point would be that we're trying to reduce server load and this wouldn't do that.

I don't think there is any live-data implementation that reduces server load. That's probably the simplest blocker to implementing it in the near term.

To expand on the discussion about what tech should be used to get the data: there are 4 main approaches to getting live data, all of which would require REST API work to support:

1. **Regular polling** - this strategy requires creating a lot of network traffic for every active user. Generally not considered a good approach for performance reasons.
2. **Long polling** - has a number of weaknesses, some of which are summarized well here: https://stackoverflow.com/a/23108363. Essentially, we would have to tell the Node server to keep multiple open requests to the API server for _every member using the site_.
3. **websockets** - when we were working on member to member messages a few years ago, websockets were the best option, and if a product team wanted live data in a new feature, I would recommend they look into setting it up that way.
4. **HTTP/2 Server Side Events** - doesn't have great support, would require major server work on the REST API and Node server.

However, I don't get the sense that live UI updating is a big product requirement for any current work. If and when it is, I think it will need to involve a full stack effort beyond the scope of the web platform, requiring significant REST API server work. Live data support is very complex and we're not going to be able to do it this year in a general way.

**Comment by willh-meetup** [comment]
> I agree that server level caching is likely out for Q3 - at least the sort thats our responsibility. Other API consumers will want caching server side caching. If we imagine having a highly cached API server, how much time would we serve by having a mup-web local cache?

@mattkime @mmcgahan I think API-caching will be a thing that we want to figure out in Q3, but maybe not implement. I think the web platform team definitely needs to be involved in figuring that out, but we will need to coordinate with the rest of the platform teams. There's also a tracking component here. If we want to track all the things, but we're not talking to the server... how do we want to do it?

>Implementing the platform as a well-defined set of modules will make all future development simpler, as well as give us the opportunity to genuinely open-source some of the more polished modules that we write - we can take the time to properly document them and set them up as packages in npm, and make a genuine effort to share what we've learned with the wider community.

Yeah, I really like this and think it's a good goal. What we do need is a metric to measure success against, and a really clear story to tell about why we think this is the way to get to that. 

>>We could do another pass at alerting / monitoring

>I'd like to know more specifics about this before putting a timeframe on it. We can and should put together a document describing our ideal alerting/monitoring behavior, but it's one of those things where it could take anywhere from 1 to 5 sprints to assemble different tools

Agree that we need to define this better and come up with what *done* looks like (or, what we think it might look like)

>>Blue / green deployment

>My impression at the moment is that this will be partially dependent on other teams, which makes it hard to timebox. Definitely something we could work on, but I'm cautious about marking it as a goal, particularly if it ended up requiring significant changes to our deployment infrastructure.

I think we need to do some investigation here, but I don't think that coordination with other teams disqualifies something from being a goal. We do need to *talk* to teams and plan, but that needs to happen anyways.

**Comment by mmcgahan** [comment]
Here's the H2 plan that I would offer. I wrote this up earlier this week as though it would be presented to Yvette or the exec team, so it includes a bit more high-level justification, but hopefully that helps clarify priorities:

---

## Overview

> First make it work, then make it right, and, finally, make it fast.  
> -- Kent Beck

The web replatform has achieved its 'phase one' goals and is now the primary framework for all new user-facing web experiences. It is stable and scalable, integrating with the rest of the Meetup ecosystem through standardized APIs and deployment workflows. It applies carefully-considered programming design patterns to modern best-of-breed open source programming libraries, enabling fast product iterations and predictable data management. It is 'fun', well-tested, and adaptable. It works.

From a pure code perspective, it is also well on its way to being 'right'. Documentation and testing is already much more extensive than anything ever produced for the (significantly larger) legacy codebase in Meetup Classic (chapstick), and the architecture of the platform makes it easier to determine where to start, how to proceed, and when to mark work as 'finished'. We are also leveraging a development community many times larger than the Meetup engineering team through extensive use of open source, building on the experience of experts in the programming community and allowing the web engineering team to participate in the industry more deeply and broadly than ever before.

So: _where do we go from here?_

This document outlines a plan to realize the potential of the platform, and specifies the relevant goals and resources needed to achieve them.

## Phase Two - consolidation and clarification

Modern software has to keep moving forward in order to avoid sliding backward. Dependencies need to stay up to date and implementations need to be refined to make them more adaptable and robust in their expected application to unpredictable future needs. There is no 'end point', although there are milestones that count as genuine achievements.

The more challenging and interesting next steps involve consolidating and strengthening the platform toolset in order to produce an architecture that has a better chance of thriving as the Meetup product adapts to a dramatically changing market. In practice, this means:

1. Reducing the 'surface area' of library code to appropriately constrain its public interface.
2. Decoupling unrelated code and organizing it better into standalone modules and packages.
3. Improving documentation and testing
4. Ideally, releasing modules more formally as open source software that can raise Meetup's engineering profile and provide opportunities to engage with the programming community more widely.

### Expected benefit

The engineering benefit of this effort will be felt every time a developer opens their code editor or fires up the development server - it will support a clearer, shared mental model of the application and ensure that the right tool is applied to the right problem at the right time more predictably, and that mistakes are more easily identifiable earlier in the dev cycle. It will make us faster.

The engineering benefit naturally supports better product development, shorter launch cycles, and reduced product resources dedicated to maintenance.

The more intangible benefit of creating better open source software should be felt in recruitment, where higher-quality engineers will be more likely to see and be interested in the technical opportunities at Meetup, as well as providing a strong professional development opportunity for current engineers.

### Resources

The web platform team will need to reshuffle in order to consolidate effectively around the needs of the platform consumers - the product teams building the Meetup web application.

I recommend that we invert our current configuration that assigns a 'point person' from the the web platform team to support each product team, and instead designate an engineer from each product team to be the point person supporting the work of the web platform. Now that a number of product teams have had a fair amount of experience developing on the platform, I believe there are engineers who would be able to effectively cross-pollinate and provide guidance toward aligning the platform with product needs.

The product 'point person' would be responsible for

1. Implementing platform updates relevant to their product team's needs.
2. Code review of any web platform work regardless of product relevance.
3. Participating in a weekly web platform meeting to steer the priorities of platform development.
4. Integrating web platform updates back into the web application.
5. Onboarding new product team engineers with relevant aspects of the platform.
6. Supporting platform documentation.

This team organization would allow the platform to become more of a shared/distributed responsibility across the web engineering team and help ensure that knowledge about the web platform is not concentrated in a small number of individuals. The organization also means that the dedicated 'Web Platform Team' might only consist of one or two engineers who can provide architectural guidance and maintain the velocity of platform development beyond the specific product-determined requirements.

### Timeline: 6 months

- 3-4 sprints: Clearly define & implement the public interface of all `meetup-web-platform` modules (#281)
- 2-3 sprints: Retire the `web-platform-starter` repo and fully document shared application patterns
- 2-3 sprints: Complete unit testing coverage (90+%) and improve static typing coverage, including exportable type definitions
- 1-2 sprints: Document and publicly release OSS packages
- remainder: Product support and planning for Phase Three

## Phase Three - Building the future

Current cutting edge web technologies enable application performance on par with native, although server, browser, and device support is still in its early stages. To build a web application that can operate as smoothly and reliably as a native app, we must start integrating these new technologies using a 'Progressive/Responsive Web Application' approach (#274). Although the timeline and resources needed to effectively implement these tools is less clear at the moment, the implementation is necessary to get from where we are to where we want to be both in terms of engineering and product.

The web platform codebase currently supports solid 2016-era applications and is almost entirely unencumbered by legacy support. It must now adopt future-oriented technologies like service workers and multiplexed HTTP/2 data connections, which will require careful investigation and experimentation. In order to enable highly-responsive web application UIs exemplified by Twitter, Facebook, and Google, we must leverage new technology or risk being stuck with great product ambitions that cannot be supported across the range of devices and network conditions that Meetup (and the Internet more generally) will be consumed from over the next few years.

### Resources

As we start integrating new technologies, we will need to develop expertise. This means making room for product-independent development, with a 'pure tech' mandate for specific individuals. Although moving these engineers out of direct product development will slow down the output of individual product teams, this work is clearly aimed at systematically improving product performance across the board - 'a rising tide lifts all boats', etc.

Specifically, we need

1. A server-side expert who can implement HTTP/2 and define a Server Push-friendly build.
2. A service worker expert who can design and implement reliable and comprehensive offline-first behavior.
3. An app-state management (Redux) expert who can define and implement clear APIs for wrangling the required asynchronous data sources in an elegant way.

These roles can be overlapping, but they shouldn't all be allocated to the same person, and they must be held by experienced developers who have a strong ability to design complex (not complicated) systems.

### Timeline: ~6 months

- 2-3? sprints: major caching update
- 3-4 sprints: Offline-first support and documentation
- 1-2 sprints: App server HTTP/2 implementation and feature planning
- 2-3? sprints: HTTP/2 Server Push implementation
- 2-3? sprints: HTTP/2 Server Side Events implementation
- 2? sprints: Build routine optimization
- 1-2 sprints: Documentation and public releases of OSS packages

## Summary

The web platform is well past being a minimally-viable product and is now well-established to support modern web development. However, its primary value is the future it enables, and this document outlines the steps necessary to realize that value. In the near term, we need to review the system and consolidate features to make it more coherent and long-term maintainable. From there, we will be able to add forward-looking features that provide both general and specific performance improvements to product development and application users.

This effort will not require a large group of dedicated engineers, but it will require distributing and organizing development efforts in a structured way. If implemented, I expect that this plan will yield a web platform that is robust and powerful enough to drive Meetup well past 2020 with an adaptable, well-designed foundation.

**Comment by mmcgahan** [comment]
>> Blue / green deployment

> I think we need to do some investigation here, but I don't think that coordination with other teams disqualifies something from being a goal

Definitely. I'm cautious about making the actual implementation of blue/green deployment a Q3 goal because of how slow it can be to coordinate infrastructure/deployment work with other teams, but we'll definitely continue working on it - making it an H2 goal seems more reasonable, or "coordinate with other teams to determine infrastructure requirements for blue/green deployments" as a more controllable Q3 goal that is less contingent on another team's availability

**Comment by willh-meetup** [comment]
> Definitely. I'm cautious about making the actual implementation of blue/green deployment a Q3 goal because of how slow it can be to coordinate infrastructure/deployment work with other teams, but we'll definitely continue working on it - making it an H2 goal seems more reasonable, or "coordinate with other teams to determine infrastructure requirements for blue/green deployments" as a more controllable Q3 goal that is less contingent on another team's availability

Yep, understood. I'm talking about coordinating with teams in the next few weeks so we can try and get a size and scope before we promise anything :).

**Comment by mattkime** [comment]
@mmcgahan I like the direction and plan but would like to see it broken down into smaller pieces to fit within Q3 and along side other work.

Is there a priority to which modules get broken out first?

**Comment by mmcgahan** [comment]
The details of the module breakdown is in #281. If we need a Q3-specific plan as opposed to H2, we can aim to complete the module refactoring and eliminating the starter repo - that seems like a good, ambitious goal for 3 months of work alongside product support.

There are a few module dependencies that would determine some dev priorities, but most of those are related to the Hapi plugin work, and are detailed in the Google Doc linked from #281

**Comment by mmcgahan** [comment]
'Future work' for 2018 will remain in the mix for 2018 planning - the remainder has mostly been sorted out
